crates/stages/src/stages/hashing_storage.rs

```diff
@@ -0,0 +1,491 @@
+                        accounts
+                    },
+                )
+                .into_iter()
+                // iterate over plain state and get newest storage value.
+                // Assumption we are okay with is that plain state represent
+                // `previous_stage_progress` state.
+                .map(|(address, storage)| {
+                    storage
+                        .into_iter()
+                        .map(|(key, val)| {
+                            plain_storage
+                                .seek_by_key_subkey(address, key)
+                                .map(|ret| (key, (val, ret.map(|e| e.value))))
+                        })
+                        .collect::<Result<BTreeMap<_, _>, _>>()
+                        .map(|storage| (address, storage))
+                })
+                .collect::<Result<Vec<_>, _>>()?
+                .into_iter()
+                // Hash the address and key and apply them to HashedStorage (if Storage is None
+                // just remove it);
+                .try_for_each(|(address, storage)| {
+                    let hashed_address = keccak256(address);
+                    storage.into_iter().try_for_each(
+                        |(key, (old_val, new_val))| -> Result<(), StageError> {
+                            let key = keccak256(key);
+                            tx.delete::<tables::HashedStorage>(
+                                hashed_address,
+                                Some(StorageEntry { key, value: old_val }),
+                            )?;
+                            if let Some(value) = new_val {
+                                let val = StorageEntry { key, value };
+                                tx.put::<tables::HashedStorage>(hashed_address, val)?
+                            }
+                            Ok(())
+                        },
+                    )
+                })?;
+        }
+
+        info!(target: "sync::stages::hashing_storage", "Stage finished");
+        Ok(ExecOutput { stage_progress: input.previous_stage_progress(), done: true })
+    }
+
+    /// Unwind the stage.
+    async fn unwind(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, StageError> {
+        let from_transition_rev = tx.get_block_transition(input.unwind_to)? + 1;
+        let to_transition_rev = tx.get_block_transition(input.stage_progress)? + 1;
+
+        let mut hashed_storage = tx.cursor_dup_write::<tables::HashedStorage>()?;
+
+        // Aggregate all transition changesets and and make list of account that have been changed.
+        tx.cursor_read::<tables::StorageChangeSet>()?
+            .walk((from_transition_rev, H160::zero()).into())?
+            .take_while(|res| {
+                res.as_ref()
+                    .map(|(TransitionIdAddress((k, _)), _)| *k < to_transition_rev)
+                    .unwrap_or_default()
+            })
```

Comment:
Would like to figure out a good helper for walking+take while, maybe like a range(..) helper generic over the key?

Reply:
it will be benefitial! This pattern is used in most of the stages.